### PR TITLE
feat: Add retrieve online documents v2 method into  pgvector 

### LINKED ISSUE #5115: Update Elastic Search, QDrant, and PGVector to `retrieve_online_documents_v2` method
**Is your feature request related to a problem? Please describe.**
Update Elastic Search, QDrant, and PGVector to `retrieve_online_documents_v2` method, as it returns broader features.

**Describe the solution you'd like**
V2 is the better method.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by YassinNouh21**
# What this PR does / why we need it:

This PR enhances the PostgreSQL online store to support **hybrid search capabilities**, combining both vector similarity search and full-text search.  

Specifically:
- Introduces the ability to perform hybrid queries using both embeddings and keyword-based search.
- Extends `retrieve_online_documents_v2` function to handle vector-only, text-only, and hybrid cases gracefully.
- Improves feature retrieval by dynamically selecting features based on query type (distance, text_rank).
- Adds comprehensive integration tests to validate:
  - Vector similarity search (L2 and cosine distance)
  - Full-text search
  - Hybrid search (vector + text)
  - Edge cases (non-matching queries, category filtering)

This update supports the broader goal of enabling more intelligent, contextual document retrieval in Feast's online stores.

# Which issue(s) this PR fixes:

Fixes #5115  
Part of the roadmap to **Introduce Feast NLP/LLM Add-On**, enabling advanced search capabilities in vector databases.

# Misc

### CODE REVIEW

File: sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py [hunk 1]
```diff
@@ -456,6 +480,267 @@ def retrieve_online_documents(
...
+        distance_metric = distance_metric or "L2"
+
+        if distance_metric not in SUPPORTED_DISTANCE_METRICS_DICT:
+            raise ValueError(
+                f"Distance metric {distance_metric} is not supported. Supported distance metrics are {SUPPORTED_DISTANCE_METRICS_DICT.keys()}"
+            )
+
+        distance_metric_sql = SUPPORTED_DISTANCE_METRICS_DICT[distance_metric]
+
+        string_fields = []
```
Comment (ntkathole) [inline sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py:529]:
```suggestion
        string_fields = [
        feature.name
        for feature in table.features
        if feature.dtype.to_value_type().value == 2 and feature.name in requested_features
    ]
```

File: sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py [hunk 2]
```diff
@@ -456,6 +480,267 @@ def retrieve_online_documents(
...
+                    INNER JOIN text_matches t2 ON t1.entity_key = t2.entity_key
+                    WHERE t1.feature_name = ANY(%s)
+                    ORDER BY t2.text_rank DESC
+                    """
+                ).format(
+                    table_name=sql.Identifier(table_name),
+                    top_k=sql.Literal(top_k),
+                )
+
+                cur.execute(
```
Comment (ntkathole) [inline sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py:626]: cur.execute and cur.fetchall() is repeated in all conditions. 

if hybrid search, params = [embedding, tsquery_str, string_fields, tsquery_str]
if vector search, params = [embedding],
.....

cur.execute(query, params)
rows = cur.fetchall()

File: sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py [hunk 3]
```diff
@@ -456,6 +480,267 @@ def retrieve_online_documents(
...
+                if distance is not None:
+                    entities_dict[key]["vector_distance"] = min(
+                        entities_dict[key]["vector_distance"], float(distance)
+                    )
+                if text_rank is not None:
+                    entities_dict[key]["text_rank"] = max(
+                        entities_dict[key]["text_rank"], float(text_rank)
+                    )
+
+            if embedding is not None and query_string is not None:
```
Comment (ntkathole) [inline sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py:691]: can this be simplified ?
```
def sort_key(item: Dict[str, Any]) -> float:
            return item["vector_distance"] if embedding else item["text_rank"]
```

**Comment by YassinNouh21** [comment]
@ntkathole  can u take a quick look

File: sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py [hunk 4]
```diff
@@ -308,6 +316,11 @@ def update(
                 else:
                     # keep the vector_value_type as BYTEA if pgvector is not enabled, to maintain compatibility
                     vector_value_type = "BYTEA"
+
+                has_string_features = any(
```
Comment (franciscojavierarceo) [inline sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py:320]: Maybe there's a more explicit way to handle this? Feels like this could be cleaner.

Comment (YassinNouh21) [reply sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py:320]: do u think this will be a better version 

```
has_string_features = any(
                    f.dtype.to_value_type() == ValueType.STRING 
                    for f in table.features
                )
```

Comment (franciscojavierarceo) [reply sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py:320]: yes!

**Comment by YassinNouh21** [comment]
@franciscojavierarceo  I think the reson behind the failed of the ci is this 
at this file in sdk/python/tests/integration/online_store/test_universal_online.py
```
 sdk/python/tests/integration/online_store/test_universal_online.py
         # writes to online store via datasource (dataframe_source) materialization
         fs.materialize(
-            start_date=datetime.datetime.now() - timedelta(hours=12),
+            start_date=datetime.now() - timedelta(hours=12),
             end_date=_utc_now(),
         )
```

File: sdk/python/tests/integration/online_store/test_universal_online.py [hunk 5]
```diff
@@ -1,8 +1,8 @@
-import datetime
 import os
+import random
 import time
 import unittest
-from datetime import timedelta
+from datetime import datetime, timedelta
```
Comment (franciscojavierarceo) [inline sdk/python/tests/integration/online_store/test_universal_online.py:5]: i believe this is what broke the integration tests

it's unfortunate naming but previously we imported `datetime` the package and now you've imported the `datetime` module from `datetime` the package, which is what leads to the issue.

**Review by franciscojavierarceo — COMMENTED** [review]
thanks for this! need to change the import or update the integration test

**Review by franciscojavierarceo — APPROVED** [review]
🚀🚀🚀
